producer/data/ref_counted.go

```diff
@@ -0,0 +1,103 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package data
+
+import (
+	"sync"
+
+	"github.com/m3db/m3msg/producer"
+
+	"go.uber.org/atomic"
+)
+
+// OnFinalizeFn will be called when the data is being finalized.
+type OnFinalizeFn func(d producer.RefCountedData)
+
+type refCountedData struct {
+	sync.RWMutex
+
+	rc       *atomic.Int32
+	d        producer.Data
+	fn       OnFinalizeFn
+	isClosed bool
+}
+
+// NewRefCountedData creates RefCountedData.
+func NewRefCountedData(data producer.Data, fn OnFinalizeFn) producer.RefCountedData {
+	return &refCountedData{
+		rc:       atomic.NewInt32(0),
```

Comment:
The more common case is to inc ref after creation? Feels more readable to have IncRef/DecRef calls paired in the code. In this case the writer will be using incref and decref to manage the lifecycle for the ref count, the buffer would only directly drop if needed.

Reply:
I usually think that whoever creates it always have a reference to it and thus should always have a positive refcount. A refcount of 0 means no one owns it and it's eligible for GC. Granted you can immediately IncRef after the ctor returns, so this is more on the semantics side.

Reply:
I feel it makes the debugging easier if you can find a inc for every dec or at least they should be close, in this case the dec can't happen in buffer because the it needs to wait until the writer to do a few incs before calling the dec for the ref counted from constructor.